File: app/typescript/src/domain/production-plan/allocation.repository.ts
import { PrismaClient } from '@prisma/client'

// 引当区分
export const AllocationCategory = {
  INVENTORY: 'INVENTORY', // 在庫
  PURCHASE_REMAIN: 'PURCHASE_REMAIN', // 発注残
  MANUFACTURING_REMAIN: 'MANUFACTURING_REMAIN', // 製造残
} as const

export type AllocationCategoryValue =
  (typeof AllocationCategory)[keyof typeof AllocationCategory]

// 引当情報の入力型
export interface CreateAllocationInput {
  requirementId: number
  allocationCategory: AllocationCategoryValue
  orderId?: number
  allocationDate: Date
  allocatedQuantity: number
  locationCode: string
}

// 引当情報の出力型
export interface Allocation {
  id: number
  requirementId: number
  allocationCategory: string
  orderId: number | null
  allocationDate: Date
  allocatedQuantity: number
  locationCode: string
  createdAt: Date
  updatedAt: Date
}

export class AllocationRepository {
  constructor(private prisma: PrismaClient) {}

  async create(input: CreateAllocationInput): Promise<Allocation> {
    const result = await this.prisma.$queryRaw<Allocation[]>`
      INSERT INTO allocations (
        requirement_id, allocation_category, order_id,
        allocation_date, allocated_quantity, location_code
      )
      VALUES (
        ${input.requirementId}, ${input.allocationCategory}::allocation_category,
        ${input.orderId ?? null}, ${input.allocationDate},
        ${input.allocatedQuantity}, ${input.locationCode}
      )
      RETURNING
        id,
        requirement_id as "requirementId",
        allocation_category as "allocationCategory",
        order_id as "orderId",
        allocation_date as "allocationDate",
        allocated_quantity as "allocatedQuantity",
        location_code as "locationCode",
        created_at as "createdAt",
        updated_at as "updatedAt"
    `
    return result[0]
  }

  async findByRequirementId(requirementId: number): Promise<Allocation[]> {
    const result = await this.prisma.$queryRaw<Allocation[]>`
      SELECT
        id,
        requirement_id as "requirementId",
        allocation_category as "allocationCategory",
        order_id as "orderId",
        allocation_date as "allocationDate",
        allocated_quantity as "allocatedQuantity",
        location_code as "locationCode",
        created_at as "createdAt",
        updated_at as "updatedAt"
      FROM allocations
      WHERE requirement_id = ${requirementId}
      ORDER BY allocation_date
    `
    return result
  }

  async findByCategory(category: AllocationCategoryValue): Promise<Allocation[]> {
    const result = await this.prisma.$queryRaw<Allocation[]>`
      SELECT
        id,
        requirement_id as "requirementId",
        allocation_category as "allocationCategory",
        order_id as "orderId",
        allocation_date as "allocationDate",
        allocated_quantity as "allocatedQuantity",
        location_code as "locationCode",
        created_at as "createdAt",
        updated_at as "updatedAt"
      FROM allocations
      WHERE allocation_category = ${category}::allocation_category
      ORDER BY allocation_date
    `
    return result
  }
}

File: app/csharp/src/ProductionManagement.Infrastructure/Persistence/Repositories/AllocationRepository.cs
using Dapper;
using Npgsql;
using ProductionManagement.Application.Port.Out;
using ProductionManagement.Domain.Models.Plan;
using ProductionManagement.Infrastructure.Persistence.TypeHandlers;

namespace ProductionManagement.Infrastructure.Persistence.Repositories;

/// <summary>
/// 引当情報リポジトリ実装
/// </summary>
public class AllocationRepository : IAllocationRepository
{
    private readonly string _connectionString;

    static AllocationRepository()
    {
        SqlMapper.AddTypeHandler(new AllocationTypeTypeHandler());
        SqlMapper.AddTypeHandler(new DateOnlyTypeHandler());
    }

    public AllocationRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    public async Task SaveAsync(Allocation allocation)
    {
        const string sql = """
            INSERT INTO "引当情報" (
                "所要ID", "引当区分", "オーダID", "引当日", "引当数量", "場所コード"
            ) VALUES (
                @RequirementId, @AllocationType::引当区分, @OrderId, @AllocationDate, @AllocatedQuantity, @LocationCode
            )
            RETURNING "ID"
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        allocation.Id = await connection.ExecuteScalarAsync<int>(sql, new
        {
            allocation.RequirementId,
            AllocationType = allocation.AllocationType.GetDisplayName(),
            allocation.OrderId,
            allocation.AllocationDate,
            allocation.AllocatedQuantity,
            allocation.LocationCode
        });
    }

    public async Task<IReadOnlyList<Allocation>> FindByRequirementIdAsync(int requirementId)
    {
        const string sql = """
            SELECT
                "ID" as Id,
                "所要ID" as RequirementId,
                "引当区分"::TEXT as AllocationTypeValue,
                "オーダID" as OrderId,
                "引当日" as AllocationDate,
                "引当数量" as AllocatedQuantity,
                "場所コード" as LocationCode,
                "作成日時" as CreatedAt,
                "更新日時" as UpdatedAt
            FROM "引当情報"
            WHERE "所要ID" = @RequirementId
            ORDER BY "引当日"
            """;

        await using var connection = new NpgsqlConnection(_connectionString);
        var result = await connection.QueryAsync<Allocation>(sql, new { RequirementId = requirementId });
        return result.ToList();
    }

    public async Task DeleteAllAsync()
    {
        const string sql = """DELETE FROM "引当情報" """;

        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.ExecuteAsync(sql);
    }
}
